### PR TITLE
DO-290: Added implementation of member registration & generate registration url API

### PR DESCRIPTION
**Pull request by sandeep-pepcus**
Confluence Document Link : 
- https://pepcus.atlassian.net/wiki/spaces/DO/pages/70746113/Generate+Registration+URL+API
- https://pepcus.atlassian.net/wiki/spaces/DO/pages/68714497/Member+Registration+API

Note : I have added new environment variable to get host info. Need to modify value at respective env : 
- application.host = qa.listanyone.com

In addition to set temporary user need to execute following queries at respective environment.
**Step 1** : Add role for temporary login
insert into core.role (name) values ('TEMPORARY')

**Step 2** : Add new temporary user for registration for each tenant (username = tenantName.reg)

insert into core.login_info (username, password, created_date, created_by, modified_date, modified_by)
values ('pyc.reg', '$2a$10$zThole9oNbdflR1kBoPs7epL57Prw6fIMzlECjm3RGhBr0Oy2hK8u','2020-06-08 19:31:25.797','admin','2020-06-08 19:31:25.797', 'admin' )

**Step 3** : Add entry for newly created user in tenant_member to associate with role & tenant

insert into core.tenant_member (tenant_id, role_id, login_id, is_active) values (<tenant_id>, <newly_temporary_role_id>, <newly_login_id>, true)

### CODE REVIEW

File: src/main/java/com/pepcus/directoryonline/api/controller/MemberController.java [hunk 1]
```diff
@@ -258,5 +262,30 @@ public MessageModel updateCredential(@PathVariable Integer id, @PathVariable Int
...
+   * Endpoint to register member 
+   * 
+   * @param registrationKey
+   * @param member
+   * @param response
+   * @return
+   * @throws IOException
+   */
+  @RequestMapping(method = RequestMethod.POST)
+  public ResponseEntity<Member> registerMember(
```
Comment (pepcus-ankit) [inline src/main/java/com/pepcus/directoryonline/api/controller/MemberController.java:276]: rename this to createMember

File: src/main/java/com/pepcus/directoryonline/api/controller/MemberController.java [hunk 2]
```diff
@@ -258,5 +262,30 @@ public MessageModel updateCredential(@PathVariable Integer id, @PathVariable Int
...
+      @RequestHeader(value = "registrationKey", required = true) String registrationKey,
+      @RequestBody Member member,
+      HttpServletResponse response) throws IOException {
+
+    Member res = memberService.registerMember(registrationKey, member);
+    
+    // Set token in cookie and send it to response
+    Cookie cookie = new Cookie("user-token", res.getToken());
```
Comment (pepcus-ankit) [inline src/main/java/com/pepcus/directoryonline/api/controller/MemberController.java:284]: Add a common utility to generate create cookie takes key & value.

File: src/main/java/com/pepcus/directoryonline/api/controller/TenantLaunchController.java [hunk 3]
```diff
@@ -39,4 +59,28 @@
...
+  @PreAuthorize("hasPermission('launch', 'SYS_ADMIN')")
+  @PostMapping("/generate-registration-url")
+  public ResponseEntity<UserRegistrationModel> getRegistrationUrl(
+      @RequestBody UserRegistrationModel userRegistrationModel, HttpServletRequest request) {
+    String tenantKey = APIRequestUtils.getTenantKeyFromRequest();
+    String protocol = request.getScheme().concat("://");
+    String registrationPath = "/user-registration";
+
+    String token = authenticationService.generateTokenForRegistration(tenantKey, userRegistrationModel);
+    String launchUrl = protocol.concat(tenantKey).concat(ApplicationConstant.DOT).concat(applicationHost)
```
Comment (pepcus-ankit) [inline src/main/java/com/pepcus/directoryonline/api/controller/TenantLaunchController.java:79]: Instead of reading applicationHost as environment variable. We can read by request.getRequestURL()  (HttpServletRequest)

File: src/main/java/com/pepcus/directoryonline/api/entity/Member.java [hunk 4]
```diff
@@ -215,6 +215,11 @@
   
   @Transient
   private Integer parentId;
+  
+  @JsonIgnore
+  @Transient
+  private String token;
```
Comment (pepcus-ankit) [inline src/main/java/com/pepcus/directoryonline/api/entity/Member.java:221]: Why do we need this. We are returning token from service layer.

File: src/main/java/com/pepcus/directoryonline/api/service/MemberService.java [hunk 5]
```diff
@@ -323,4 +344,90 @@ public void setTopFields(List<String> topFields, Member memberObj, Member member
...
+      throw ApplicationException.createAccessDeniedError(APIErrorCodes.ACCESS_DENIED);
+    }
+
+    String tenantKey = APIRequestUtils.getTenantKeyFromRequest();
+    Tenant tenant = tenantService.getTenantByKey(tenantKey);
+
+    memberValidator.validateRegisterMemberRequest(tenantKey, member);
+
+    // set member creation info
+    member.setDateOfBirth(getDate(getDate(member.getDateOfBirth(), DATE_PATTERN)));
```
Comment (pepcus-ankit) [inline src/main/java/com/pepcus/directoryonline/api/service/MemberService.java:367]: why getDate(getDate ??

File: src/main/java/com/pepcus/directoryonline/api/service/MemberService.java [hunk 6]
```diff
@@ -323,4 +344,90 @@ public void setTopFields(List<String> topFields, Member memberObj, Member member
...
+    member.setCreatedBy(USER_ADMIN);
```
Comment (pepcus-ankit) [inline src/main/java/com/pepcus/directoryonline/api/service/MemberService.java:368]: We should rather set the username of token used here. instead of hardcoding it as USER_ADMIN

File: src/main/java/com/pepcus/directoryonline/api/service/MemberService.java (hunk 5 above)
Comment (sandeep-pepcus) [reply src/main/java/com/pepcus/directoryonline/api/service/MemberService.java:367]: We are using Entity as model. And DOB is define as date datatype. In order to type conversion we need to change it string first and then format & again convert to date and persist.

File: src/main/java/com/pepcus/directoryonline/api/service/MemberService.java [hunk 7]
```diff
@@ -323,4 +344,90 @@ public void setTopFields(List<String> topFields, Member memberObj, Member member
...
+    
+    return true;
+  }
+  
+  /**
+   * Method to check duplicate phone number
+   * @param memberId
+   * @param phoneNumber
+   */
+  public void checkDuplicatePhonenumber(Integer memberId, String phoneNumber) {
```
Comment (pepcus-ankit) [inline src/main/java/com/pepcus/directoryonline/api/service/MemberService.java:417]: checkDuplicatePhoneNumber (case for N)

File: src/main/resources/application.properties [hunk 8]
```diff
@@ -40,4 +40,6 @@ application.system.admin.contact = IAMDUMMY
 application.sms.token = IAMDUMMY
 
 # Analytic related configurations
-application.analytics.ga.tracking.id = 
\ No newline at end of file
+application.analytics.ga.tracking.id = 
+
+application.host = qa.listanyone.com
```
Comment (pepcus-ankit) [inline src/main/resources/application.properties:45]: I think we won't need this if we read request URL from httpservletrequest.

File: src/main/java/com/pepcus/directoryonline/api/service/MemberService.java (hunk 7 above)
Comment (sandeep-pepcus) [reply src/main/java/com/pepcus/directoryonline/api/service/MemberService.java:417]: In register member API, we have only one phone number for register. So added check only for one member. For n we need to implement logic to validate n number with ids (Complex logic).

File: src/main/resources/application.properties (hunk 8 above)
Comment (sandeep-pepcus) [reply src/main/resources/application.properties]: Addressed

File: src/main/java/com/pepcus/directoryonline/api/controller/TenantLaunchController.java (hunk 3 above)
Comment (sandeep-pepcus) [reply src/main/java/com/pepcus/directoryonline/api/controller/TenantLaunchController.java]: addressed

File: src/main/java/com/pepcus/directoryonline/api/entity/Member.java (hunk 4 above)
Comment (sandeep-pepcus) [reply src/main/java/com/pepcus/directoryonline/api/entity/Member.java]: Fixed

File: src/main/java/com/pepcus/directoryonline/api/service/MemberService.java (hunk 6 above)
Comment (sandeep-pepcus) [reply src/main/java/com/pepcus/directoryonline/api/service/MemberService.java]: Addressed

File: src/main/java/com/pepcus/directoryonline/api/controller/MemberController.java (hunk 2 above)
Comment (sandeep-pepcus) [reply src/main/java/com/pepcus/directoryonline/api/controller/MemberController.java]: Addressed

File: src/main/java/com/pepcus/directoryonline/api/controller/MemberController.java (hunk 1 above)
Comment (sandeep-pepcus) [reply src/main/java/com/pepcus/directoryonline/api/controller/MemberController.java]: Changed

File: src/main/java/com/pepcus/directoryonline/api/controller/MemberController.java [hunk 9]
```diff
@@ -258,5 +261,30 @@ public MessageModel updateCredential(@PathVariable Integer id, @PathVariable Int
...
+      HttpServletResponse response) throws IOException {
+
+    String token = memberService.createMember(registrationKey, member);
+    
+    // Set token in cookie and send it to response
+    Cookie cookie = CommonUtils.createCookie("user-token", token);
+    cookie.setPath("/");
+    response.addCookie(cookie);
+    
+    return new ResponseEntity<>(member, HttpStatus.OK);
```
Comment (pepcus-ankit) [inline src/main/java/com/pepcus/directoryonline/api/controller/MemberController.java:287]: How will UI get the member id?

File: src/main/java/com/pepcus/directoryonline/api/validator/MemberValidator.java [hunk 10]
```diff
@@ -419,4 +425,30 @@ private static boolean validateUrlMemberId(Integer requestedMemberId, String url
...
+    validateValues("salutation", request.getSalutation(), dropdownMap.get(SALUTATION), true);
+    validateValues("profession",  request.getProfession(), dropdownMap.get(PROFESSION), true);
+    validate("specialization", request.getSpecialization(), expect(onlyAllowedChars, maxLength(256)));
+    validate("primaryNumber", request.getPrimaryNumber(), expect(nonEmpty, phone));
+    validate("email", request.getEmail(), expect(email, maxLength(100)));
+    validate("username", request.getUsername(), expect(nonEmpty, minLength(6), maxLength(64)));
+    CommonValidator.validateUsername("username", request.getUsername());
+    validate("password", request.getPassword(), expect(nonEmpty, minLength(6), maxLength(20)));
+    CommonValidator.validatePassword("password", request.getPassword()); 
+    memberService.checkDuplicatePhonenumber(null, request.getPrimaryNumber());
```
Comment (pepcus-ankit) [inline src/main/java/com/pepcus/directoryonline/api/validator/MemberValidator.java:450]: rename method checkDuplicatePhonenumber -> checkDuplicatePhoneNumber (with capital N)

File: src/main/resources/application.properties [hunk 11]
```diff
@@ -40,4 +40,4 @@ application.system.admin.contact = IAMDUMMY
 application.sms.token = IAMDUMMY
 
 # Analytic related configurations
-application.analytics.ga.tracking.id = 
\ No newline at end of file
+application.analytics.ga.tracking.id =
```
Comment (pepcus-ankit) [inline src/main/resources/application.properties:43]: Revert this change.

File: src/main/java/com/pepcus/directoryonline/api/controller/TenantLaunchController.java [hunk 12]
```diff
@@ -39,4 +50,27 @@
...
+  @PreAuthorize("hasPermission('launch', 'SYS_ADMIN')")
+  @PostMapping("/generate-registration-url")
+  public ResponseEntity<UserRegistrationModel> getRegistrationUrl(
+      @RequestBody UserRegistrationModel userRegistrationModel, HttpServletRequest request) {
+    String tenantKey = APIRequestUtils.getTenantKeyFromRequest();
+    String protocol = request.getScheme().concat("://");
+    String registrationPath = "/user-registration";
+
+    String token = authenticationService.generateTokenForRegistration(tenantKey, userRegistrationModel);
+    String launchUrl = protocol.concat(request.getHeader(HttpHeaders.HOST)).concat(registrationPath).concat("?p=").concat(token);
```
Comment (pepcus-ankit) [inline src/main/java/com/pepcus/directoryonline/api/controller/TenantLaunchController.java:70]: Please check if we really need to get protocol & host separately? I guess we can directly get both as single header.

File: src/main/java/com/pepcus/directoryonline/api/utils/CommonUtils.java [hunk 13]
```diff
@@ -224,4 +225,15 @@ public static boolean isMobileNumberContainsIndianFormat(String registeredMobile
...
+  
+  /**
+   * Method to create cookie with given key & value
+   * 
+   * @param key
+   * @param value
+   * @return
+   */
+  public static Cookie createCookie(String key, String value){
+    return new Cookie(key, value);
```
Comment (pepcus-ankit) [inline src/main/java/com/pepcus/directoryonline/api/utils/CommonUtils.java:237]: I think we can set the path here only. No need to set it separately everywhere.

File: src/main/java/com/pepcus/directoryonline/api/controller/MemberController.java (hunk 9 above)
Comment (sandeep-pepcus) [reply src/main/java/com/pepcus/directoryonline/api/controller/MemberController.java:285]: Member object is send in response and it contains member id.

File: src/main/java/com/pepcus/directoryonline/api/validator/MemberValidator.java (hunk 10 above)
Comment (sandeep-pepcus) [reply src/main/java/com/pepcus/directoryonline/api/validator/MemberValidator.java]: Addressed

File: src/main/resources/application.properties (hunk 11 above)
Comment (sandeep-pepcus) [reply src/main/resources/application.properties]: Reverted. After this change now TenantLaunch API will not return trackingId in response.

File: src/main/java/com/pepcus/directoryonline/api/utils/CommonUtils.java (hunk 13 above)
Comment (sandeep-pepcus) [reply src/main/java/com/pepcus/directoryonline/api/utils/CommonUtils.java]: Done

File: src/main/java/com/pepcus/directoryonline/api/controller/TenantLaunchController.java (hunk 12 above)
Comment (sandeep-pepcus) [reply src/main/java/com/pepcus/directoryonline/api/controller/TenantLaunchController.java:70]: Followings method to get server/url/uri from request. No one method will return only protocal & domain in single header. 
Examples : 
"request.getServerName()" :    localhost
"request.getRequestURI()" :     /api/v1/members
"request.getRequestURL()" :     http://localhost:8080/api/v1/members
"request.getScheme()" :           http
"request.getServerName()" :      localhost
"request.getHeader(HttpHeaders.HOST)" :   localhost:8080

File: src/main/resources/application.properties [hunk 14]
```diff
@@ -39,5 +39,3 @@ application.system.admin.contact = IAMDUMMY
 # SMS related configurations
 application.sms.token = IAMDUMMY
 
-# Analytic related configurations
-application.analytics.ga.tracking.id =
```
Comment (pepcus-ankit) [inline src/main/resources/application.properties:43]: Bhai bhai bhai... isko kyo hataya?? 

Mene changes revert karne bola tha taki ye file commit me na aaye..... Naki tracking id ko hatane

Comment (sandeep-pepcus) [reply src/main/resources/application.properties]: Bhai fix far diya. Me samjha tha ki ye bhi revert karna he shayad UI per handle kar liya hoga trackingid wala.
